numpy/core/src/multiarray/dtype_transfer.c

```diff
@@ -464,6 +464,62 @@ _strided_to_strided_contig_align_wrap_init_dest(char *dst, npy_intp dst_stride,
     }
 }
 
+
+static void
+_strided_to_strided_contig_align_wrap_init_dest_refcnt(char *dst, npy_intp dst_stride,
+                        char *src, npy_intp src_stride,
+                        npy_intp N, npy_intp src_itemsize,
+                        NpyAuxData *data)
+{
+    _align_wrap_data *d = (_align_wrap_data *)data;
+    PyArray_StridedUnaryOp *wrapped = d->wrapped,
+            *tobuffer = d->tobuffer,
+            *frombuffer = d->frombuffer;
+    npy_intp inner_src_itemsize = d->src_itemsize,
+             dst_itemsize = d->dst_itemsize;
+    NpyAuxData *wrappeddata = d->wrappeddata,
+            *todata = d->todata,
+            *fromdata = d->fromdata;
+    char *bufferin = d->bufferin, *bufferout = d->bufferout;
+    PyArray_StridedUnaryOp *_strided_to_strided = PyArray_GetStridedCopyFn(
+            0, dst_stride, sizeof(PyObject *), sizeof(PyObject *));
```

Comment:
This is the actual change/hack. Instead of just using `memset` to initialize the buffer, copy the data from the original array into the buffer, so that decref'ing can occur. I could imagine that we can go somewhere earlier to a slow direct cast loop, which calls the cast function for every single element instead of attempting to speed it up by doing it in chunks (frankly, I doubt that helps for objects that need casting anyway).